torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp

```diff
@@ -1168,32 +1172,34 @@ class TORCH_API ProcessGroupNCCL : public Backend {
   // Mutex to guard maps like devNCCLCommMap_.
   std::mutex mutex_;
 
+  static c10::once_flag initFlag_;
+
   // Heartbeat of watchdog thread.
-  std::atomic_uint64_t heartbeat_{};
+  static std::atomic_uint64_t heartbeat_;
 
   // The time interval used for deciding whether there is no watchdog heartbeat.
-  int heartbeatTimeoutInSec_;
+  static int heartbeatTimeoutInSec_;
```

Comment:
You can pack all these variables and the `std::thread ncclHeartbeatMonitorThread_` into a `HeartbeatMonitor` class. Then create a static object. And init of all these values can be moved into constructor of the `HeartbeatMonitor` class (instead of in constructor of `ProcessGroupNCCL`)